ql/src/java/org/apache/hadoop/hive/ql/queryhistory/QueryHistoryService.java

```diff
@@ -0,0 +1,248 @@
+
+    repository = createRepository(conf);
+
+    printConfigInformation();
+
+    LOG.info("QueryHistoryService created [maxBatchSize: {}, maxQueueSizeInMemoryBytes: {}]",
+        maxBatchSize, maxQueueSizeInMemoryBytes);
+  }
+
+  /**
+   * Start method actually starts the service in a sense that it initializes the repository (which might involve
+   * database operations) and starts the executor that's responsible for persisting records.
+   * @return the service instance
+   */
+  public QueryHistoryService start() {
+    repository.init(conf, new QueryHistorySchema());
+
+    try {
+      persistExecutor.submit(() -> {
+        // this session won't need tez, let's not mess with starting a TezSessionState
+        conf.setBoolVar(HiveConf.ConfVars.HIVE_CLI_TEZ_INITIALIZE_SESSION, false);
+
+        // localize this conf for the thread that will be running the repository,
+        // as iceberg integration heavily relies on the session state conf
+        SessionState session = SessionState.start(conf);
+        LOG.info("Session for QueryHistoryService started, sessionId: {}", session.getSessionId());
+      });
+    } catch (Exception e) {
+      throw new RuntimeException("Failed to start QueryHistoryService", e);
+    }
+
+    return this;
+  }
+
+  public static QueryHistoryService getInstance() {
+    if (INSTANCE == null) {
+      throw new RuntimeException("QueryHistoryService is not present when asked for the singleton instance");
+    }
+    return INSTANCE;
+  }
+
+  // keeping a global, static instance is usually considered an antipattern, even in case of using the Singleton
+  // design, this time the constraint was the ability to hook into the query processing (Driver) from somewhere the
+  // outer service (HiveServer2, QTestUtil, etc.), which otherwise would have needed passing a service instance
+  // through many layers, making the code and method signatures more complicated
+  public static void setInstance(QueryHistoryService queryHistoryService) {
+    INSTANCE = queryHistoryService;
+  }
+
+  private void printConfigInformation() {
+    if (maxBatchSize == 0) {
+      LOG.info("Query History batch size is 0: every record will be persisted synchronously after the query" +
+          " (not recommended)");
+    } else {
+      LOG.info("Query History batch size is {}: records will be persisted after every {} queries",
+          maxBatchSize, maxBatchSize);
+    }
+    if (maxQueueSizeInMemoryBytes == 0) {
+      LOG.info("Query History max queue memory byte size is 0: the service won't check the consumed memory to make " +
+          "persist decision (not recommended)");
+    } else {
+      LOG.info("Query History max queue memory byte size is {}: records will be persisted when the queue consumes " +
+              "more than {} bytes in the memory",
+          maxQueueSizeInMemoryBytes, maxQueueSizeInMemoryBytes);
+    }
+  }
+
+  protected QueryHistoryRepository createRepository(HiveConf conf) {
```

Comment:
why is it protected?

Reply:
should be private, most probably just a leftover of VisibleForTesting scenario